File: toolinfo.pl
#!/usr/bin/perl

use strict;
use warnings;
use Getopt::Long;

##This script is for finding Tool information and User information based on passed file in comm line


my $file_name;
my $user_name;
my $help;
GetOptions("filename|file=s" => \$file_name, "username|user=s" => \$user_name, "help|h" => \$help) or die("Error in comm line args\n");

my $tool_user={};

if(defined($help)) {
 ##function call
 &help();
} ##end if

if(defined($file_name) & defined($user_name)) {
	##function call
	$tool_user=&toolInfo($file_name); 
		
	print "\n";
	print "*****User details*****\n";
	print "User Name: $user_name\n";
	print "------------------------------------------\n";
	print "Tool name\t\t No.of licenses\n";
	print "------------------------------------------\n";

	while((my $key, my $value) = each(%{$tool_user})) {
		#print "$key,$value\n";
		#print "$tool_user->{'user'}\n";
		while((my $k, my $v) = each(%{$tool_user->{$key}})) {
		#	print "$k => $v\n";
			while((my $uk,my $uv) =each(%{$tool_user->{$key}->{$k}})) {
				#	print "$uk,$uv\n";
				if($uk eq $user_name) {
					print "$k \t\t $uv\n";
				} ##end if
			} ##end while
		} ##end while
	} ##end while
	print "------------------------------------------\n";
}
elsif(defined($file_name)) {	
	##function call
	$tool_user=&toolInfo($file_name);
	while((my $k, my $v) = each(%{$tool_user->{'user'}})) {
		while((my $a,my $b) = each(%{$tool_user->{'user'}->{$k}})) {
			#print "hash: $a => $b\n";
		}
	}

	print "\n";
	print "*****Tool Details*****\n";
	while((my $hk,my $hv) = each(%{$tool_user})) {
		if($hk eq 'tool') {
			while((my $key,my $value) = each(%{$tool_user->{$hk}})) {
				print "Tool Name: $key\n";
				print "No.of licenses issued: $tool_user->{'tool'}->{$key}->{'li issued'}\n"; 		
				print "No.of licenses in use: $tool_user->{'tool'}->{$key}->{'li use'}\n";
				print "--------------------------------\n";
				print "User Name\t No.of Licenses\n";
				print "--------------------------------\n";
				while((my $uk,my $uv) = each(%{$tool_user->{'user'}})) {
					#print "$uk => $uv\n";
					if($uk eq $key) {
						#print "$uk => $uv\n";	
						#print "$tool_user->{'user'}->{'XT_XCC_TIE'}->{'sdandu'}\n";
						while((my $k,my $v) =each(%{$tool_user->{'user'}->{$uk}})) {
							print "$k     \t $v\n";					
						} ##end while
					} ##end if
				} ##end while
				print "-------------------------------------\n";
			} ##end while
		} ##end if
	} ##end while
}	##end if



##function definition:-toolInfo-
sub toolInfo {
	$file_name=shift;
	
	##opening a file
	my $tool_info={};
	my @usernames=();
	my $tuser=0;
	my $toolname=' ';
	my $tool_user->{'user'}={};
	open(IN,"$file_name") || die "file not found\n";

	while(my $lines=<IN>) {
		if($lines =~ /Users\s+of\s+(\w+):/) {
			if($lines =~ /Users\s+of\s+(\w+):\s+\(Total\s+of\s+(\d+)\s+licenses\s+issued;\s+Total\s+of\s+(\d+)\s+\w+\s+in\s+use\)/) {
     		if($3 > 0) {
					$tuser=1;
				}

				$toolname=$1;
				$tool_info->{$toolname}={'li issued' => $2, 'li use' => $3};
				next;
			} ##end if 
		} ##end if
		elsif($lines =~ /(\w+)\s+.*\s+start/) {
			#print "user_name: $1\n";
			push(@usernames,$1);
			next;
		} ##end if

		if($tuser == 1){
				my %username=();
				foreach my $k (@usernames) {
					if($k eq (keys %username)) {
			 			$username{$k} += 1;
					}
					else{
						$username{$k} = 1;
					}
				} ##end foreach
				foreach my $uk (keys %username) {
				#	print "check hash: $uk => $username{$uk}\n";
				} ##end foreach
			
				$tool_user->{'user'}->{$toolname}=\%username;
		} ##end if
		@usernames=();
 	} ##end while	

  ##closing file
	close(IN);
	
	$tool_user->{'tool'}=$tool_info;
 	
	return $tool_user;
} ##end function


##function definition: -help-
sub help {
	print "$0 => This scriptis used to find infomation of tools,licenses used & finding user info\n";
	print "filename|file <Enter file name> username|user <Enter user name>  <mandatory>
				help|h <print script usage options>
				Ex: $0 -file file1.txt -user user1\n";
} ##end function

File: toolinfo.py
#!/usr/bin/python3
import argparse;
import re;

##This script is for finding Tool infomation

##function definition
def toolInfo(filename):
	#print(filename);
	tool_dict={};
	user_dict={};
	user_dict['user']={};
	names_list=[];
	tuser=0;
	toolname=' ';
	with open(filename,"r") as fobj:
		for line in fobj:
			line=line.strip();
			ttool=re.search(r'Users\s+of\s+(\w+):',line);
			tinfo=re.search(r'Users\s+of\s+(\w+):\s+\(Total\s+of\s+(\d+)\s+licenses\s+issued;\s+Total\s+of\s+(\d+)\s+\w+\s+in\s+use\)',line);
			lineflag=re.search(r'floating\s+license',line);
			uinfo=re.search(r'(\w+)\s+.*\s+start',line);
		
			if ttool is not None:
				if tinfo is not None:
					if int(tinfo.group(3)) > 0:
						tuser=1;
					else:
						tuser=0;
					
					toolname=tinfo.group(1);
					tool_dict[toolname]={'li issued': tinfo.group(2), 'li use': tinfo.group(3)};
					continue;
			elif uinfo is not None:
				names_list.append(uinfo.group(1));
				continue;	
			
			if tuser == 1:
				usernames={};
				for un in names_list:
					#print(names_list);
					if un in usernames.keys():
						usernames[un] = usernames[un] + 1;
					else:
						usernames[un] = 1;
					#for k,v in usernames.items():
						#print(k,v);
					user_dict['user'][toolname]=usernames;
					#print(user_dict);
			names_list=[];
	
	user_dict['tool']=tool_dict;
	
	return user_dict;	
##end function


if __name__=="__main__":
	parser=argparse.ArgumentParser(description='Passing file name and finding tool & user infomation');
	parser.add_argument('-filename', '-file', dest='file', help='Enter filename');
	parser.add_argument('-username', '-user', dest='user', help='Enter username');
	args=parser.parse_args();
	#print(args);	

	tool_dict={};
	
	if args.user is not None:
		tool_dict=toolInfo(args.file);   ##function call
		#print(tool_dict);		

		print("\n");
		print("******User details******");
		print("User Name:", args.user);
		print("------------------------------------------");
		print("Tool name\t\t No.of licenses");
		print("------------------------------------------");
		for key,value in tool_dict.items():
			#print(key);
			for k,v in tool_dict[key].items():
				#print(k,v);
				for uk,uv in tool_dict[key][k].items():
					if uk == args.user:
						print("{} \t\t {}".format(k,uv));
		print("------------------------------------------");	
	elif args.file is not None:
		tool_dict=toolInfo(args.file);    ##function call
		#print(tool_dict);
	
		print("\n");
		print("******Tool Details******");
		for dk,dv in tool_dict.items():
			if dk == 'tool':
				for key,value in tool_dict[dk].items():
					#print(key);
					print("Tool Name:", key);
					print("No.of licenses issued:", tool_dict['tool'][key]['li issued']);
					print("No.of licenses in use:", tool_dict['tool'][key]['li use']);
					print("---------------------------------");
					print("User Name\t No.of Licenses");
					print("---------------------------------");
					for uk,uv in tool_dict.items():
						#print(uk,uv);
						if uk == 'user':
							for tuk,tuv in tool_dict[uk].items():
								if tuk == key:
									for k,v in tool_dict[uk][tuk].items():
										print("{}     \t {}".format(k,v));
					print("----------------------------------");	
